File: ruby/grains/grains.rb
class Grains
  VALID_SQUARES = [*1..64]

  class << self
    def square(num)
      validate_square_number(num)
      2 ** (num - 1)
    end

    def total
      VALID_SQUARES.map { |num| square(num) }.sum
    end

    def validate_square_number(num)
      raise ArgumentError unless VALID_SQUARES.include?(num)
    end
  end
end

File: typescript/grains/grains.ts
export const square = (num: number): bigint => {
  assertValidSquare(num);

  return BigInt(2 ** (num - 1));
}

export const total = () => {
  return chessboardIndexes().reduce(
    (total: bigint, num: number): bigint => total + square(num),
    BigInt(0)
  );
}

function assertValidSquare(num: number): void {
  if (num < 1 || num > 64) throw new Error("Invalid square");
}

function chessboardIndexes() {
  return [...Array(64).keys()].map(x => x + 1);
}
